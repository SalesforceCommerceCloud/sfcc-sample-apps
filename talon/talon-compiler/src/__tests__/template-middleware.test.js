const { generateHTML, templateMiddleware } = require('../template-middleware');
const { startTestContext, endTestContext } = require('test-talon-context');
const resourceService = require('../resources/resource-service');
const { getOutputConfigs } = require('talon-common');

jest.mock('../resources/resource-service');
jest.mock('../utils/observable-folder-hash');

const versionKey = '1';

beforeEach(() => {
    return startTestContext({ versionKey });
});

afterEach(() => {
    endTestContext();
});

describe('template-middleware', () => {
    const testSourceNonce = "T3stVa1ue";

    describe('generateHTML', () => {
        const testPath = '/my-page';

        getOutputConfigs().forEach(({ mode }) => {
            it(`generates HTML in ${mode} mode`, async () => {
                return generateHTML(mode, testPath).then(html => {
                    expect(html).toMatchSnapshot();
                });
            });
        });

        it('generates HTML with script-src nonce', () => {
            return generateHTML('dev', testPath, testSourceNonce).then(html => {
                expect(html.match(/nonce="T3stVa1ue"/g)).toHaveLength(2);
                expect(html).toMatchSnapshot();
            });
        });

        it('includes UIDs', async () => {
            const mode = 'dev';
            resourceService.get.mockImplementation(descriptor => {
                const uids = {
                    dev: `[${mode} uid for ${descriptor}]`
                };
                return Promise.resolve({ descriptor, uids });
            });

            return generateHTML(mode, '/').then(html => {
                expect(html).toMatchSnapshot();
            });
        });
    });

    describe('templateMiddleware', () => {
        it('send HTML', () => {
            const middleware = templateMiddleware();

            const req = {
                query: {
                    mode: 'dev'
                },
                originalUrl: '/my-page?a=1'
            };

            const res = {
                send: jest.fn(),
                locals : {
                    nonce: testSourceNonce
                }
            };

            function next(err) {
                throw err;
            }

            return middleware(req, res, next).then(() => {
                expect(res.send).toHaveBeenCalledTimes(1);
                expect(res.send.mock.calls[0][0]).toMatchSnapshot();
            });
        });
    });
});