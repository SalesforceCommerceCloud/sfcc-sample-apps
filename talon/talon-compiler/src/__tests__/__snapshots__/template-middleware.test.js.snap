// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template-middleware generateHTML generates HTML in compat mode 1`] = `
"<script src=\\"/basePath/talon/framework/latest/compat/talon.js\\"></script>

    <script type=\\"text/javascript\\" >
    Talon.configProvider.register({
        getBasePath: function getBasePath() {
            return \\"/basePath\\";
        },
        getMode: function getMode() {
            return \\"compat\\";
        },
        getLocale: function getLocale() {
            return \\"en_US\\";
        },
        getUser: function getUser() {
            return {\\"isGuest\\":true};
        }
    });
    Talon.themeService.setTheme({\\"themeLayouts\\":{\\"inner\\":{\\"view\\":\\"themeLayoutInner\\"},\\"default\\":{\\"view\\":\\"themeLayoutDefault\\"}}});
    Talon.themeService.setViewToThemeLayoutMap({\\"home\\":\\"themeLayoutInner\\",\\"myPage\\":\\"themeLayoutInner\\",\\"error\\":\\"themeLayoutDefault\\"});
    Talon.routingService.registerRoutes([{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"isRoot\\":true,\\"view\\":\\"home\\"},{\\"name\\":\\"my-page\\",\\"path\\":\\"/my-page\\",\\"view\\":\\"myPage\\"},{\\"name\\":\\"error\\",\\"path\\":\\"/error\\",\\"isDefault\\":true,\\"view\\":\\"error\\"}]);
    Talon.brandingService.setBrandingProperties({});
    
</script>
<script src=\\"/basePath/talon/view/latest/compat/en_US/themeLayoutInner.js\\"></script>
<script src=\\"/basePath/talon/view/latest/compat/en_US/myPage.js\\"></script>

<script type=\\"text/javascript\\" >
    Talon.componentService.createElement(\\"talon/app\\").then(function(element) {
        document.getElementById(\\"main\\").appendChild(element);
    });
</script>

Base path: /basePath
Version key: 1
Source nonce: "
`;

exports[`template-middleware generateHTML generates HTML in dev mode 1`] = `
"<script src=\\"/basePath/talon/framework/latest/dev/talon.js\\"></script>

    <script type=\\"text/javascript\\" >
    Talon.configProvider.register({
        getBasePath: function getBasePath() {
            return \\"/basePath\\";
        },
        getMode: function getMode() {
            return \\"dev\\";
        },
        getLocale: function getLocale() {
            return \\"en_US\\";
        },
        getUser: function getUser() {
            return {\\"isGuest\\":true};
        }
    });
    Talon.themeService.setTheme({\\"themeLayouts\\":{\\"inner\\":{\\"view\\":\\"themeLayoutInner\\"},\\"default\\":{\\"view\\":\\"themeLayoutDefault\\"}}});
    Talon.themeService.setViewToThemeLayoutMap({\\"home\\":\\"themeLayoutInner\\",\\"myPage\\":\\"themeLayoutInner\\",\\"error\\":\\"themeLayoutDefault\\"});
    Talon.routingService.registerRoutes([{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"isRoot\\":true,\\"view\\":\\"home\\"},{\\"name\\":\\"my-page\\",\\"path\\":\\"/my-page\\",\\"view\\":\\"myPage\\"},{\\"name\\":\\"error\\",\\"path\\":\\"/error\\",\\"isDefault\\":true,\\"view\\":\\"error\\"}]);
    Talon.brandingService.setBrandingProperties({});
    
</script>
<script src=\\"/basePath/talon/view/latest/dev/en_US/themeLayoutInner.js\\"></script>
<script src=\\"/basePath/talon/view/latest/dev/en_US/myPage.js\\"></script>

<script type=\\"text/javascript\\" >
    Talon.componentService.createElement(\\"talon/app\\").then(function(element) {
        document.getElementById(\\"main\\").appendChild(element);
    });
</script>

Base path: /basePath
Version key: 1
Source nonce: "
`;

exports[`template-middleware generateHTML generates HTML in prod mode 1`] = `
"<script src=\\"/basePath/talon/framework/latest/prod/talon.js\\"></script>

    <script type=\\"text/javascript\\" >
    Talon.configProvider.register({
        getBasePath: function getBasePath() {
            return \\"/basePath\\";
        },
        getMode: function getMode() {
            return \\"prod\\";
        },
        getLocale: function getLocale() {
            return \\"en_US\\";
        },
        getUser: function getUser() {
            return {\\"isGuest\\":true};
        }
    });
    Talon.themeService.setTheme({\\"themeLayouts\\":{\\"inner\\":{\\"view\\":\\"themeLayoutInner\\"},\\"default\\":{\\"view\\":\\"themeLayoutDefault\\"}}});
    Talon.themeService.setViewToThemeLayoutMap({\\"home\\":\\"themeLayoutInner\\",\\"myPage\\":\\"themeLayoutInner\\",\\"error\\":\\"themeLayoutDefault\\"});
    Talon.routingService.registerRoutes([{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"isRoot\\":true,\\"view\\":\\"home\\"},{\\"name\\":\\"my-page\\",\\"path\\":\\"/my-page\\",\\"view\\":\\"myPage\\"},{\\"name\\":\\"error\\",\\"path\\":\\"/error\\",\\"isDefault\\":true,\\"view\\":\\"error\\"}]);
    Talon.brandingService.setBrandingProperties({});
    
</script>
<script src=\\"/basePath/talon/view/latest/prod/en_US/themeLayoutInner.js\\"></script>
<script src=\\"/basePath/talon/view/latest/prod/en_US/myPage.js\\"></script>

<script type=\\"text/javascript\\" >
    Talon.componentService.createElement(\\"talon/app\\").then(function(element) {
        document.getElementById(\\"main\\").appendChild(element);
    });
</script>

Base path: /basePath
Version key: 1
Source nonce: "
`;

exports[`template-middleware generateHTML generates HTML in prod_compat mode 1`] = `
"<script src=\\"/basePath/talon/framework/latest/prod_compat/talon.js\\"></script>

    <script type=\\"text/javascript\\" >
    Talon.configProvider.register({
        getBasePath: function getBasePath() {
            return \\"/basePath\\";
        },
        getMode: function getMode() {
            return \\"prod_compat\\";
        },
        getLocale: function getLocale() {
            return \\"en_US\\";
        },
        getUser: function getUser() {
            return {\\"isGuest\\":true};
        }
    });
    Talon.themeService.setTheme({\\"themeLayouts\\":{\\"inner\\":{\\"view\\":\\"themeLayoutInner\\"},\\"default\\":{\\"view\\":\\"themeLayoutDefault\\"}}});
    Talon.themeService.setViewToThemeLayoutMap({\\"home\\":\\"themeLayoutInner\\",\\"myPage\\":\\"themeLayoutInner\\",\\"error\\":\\"themeLayoutDefault\\"});
    Talon.routingService.registerRoutes([{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"isRoot\\":true,\\"view\\":\\"home\\"},{\\"name\\":\\"my-page\\",\\"path\\":\\"/my-page\\",\\"view\\":\\"myPage\\"},{\\"name\\":\\"error\\",\\"path\\":\\"/error\\",\\"isDefault\\":true,\\"view\\":\\"error\\"}]);
    Talon.brandingService.setBrandingProperties({});
    
</script>
<script src=\\"/basePath/talon/view/latest/prod_compat/en_US/themeLayoutInner.js\\"></script>
<script src=\\"/basePath/talon/view/latest/prod_compat/en_US/myPage.js\\"></script>

<script type=\\"text/javascript\\" >
    Talon.componentService.createElement(\\"talon/app\\").then(function(element) {
        document.getElementById(\\"main\\").appendChild(element);
    });
</script>

Base path: /basePath
Version key: 1
Source nonce: "
`;

exports[`template-middleware generateHTML generates HTML in prod_debug mode 1`] = `
"<script src=\\"/basePath/talon/framework/latest/prod_debug/talon.js\\"></script>

    <script type=\\"text/javascript\\" >
    Talon.configProvider.register({
        getBasePath: function getBasePath() {
            return \\"/basePath\\";
        },
        getMode: function getMode() {
            return \\"prod_debug\\";
        },
        getLocale: function getLocale() {
            return \\"en_US\\";
        },
        getUser: function getUser() {
            return {\\"isGuest\\":true};
        }
    });
    Talon.themeService.setTheme({\\"themeLayouts\\":{\\"inner\\":{\\"view\\":\\"themeLayoutInner\\"},\\"default\\":{\\"view\\":\\"themeLayoutDefault\\"}}});
    Talon.themeService.setViewToThemeLayoutMap({\\"home\\":\\"themeLayoutInner\\",\\"myPage\\":\\"themeLayoutInner\\",\\"error\\":\\"themeLayoutDefault\\"});
    Talon.routingService.registerRoutes([{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"isRoot\\":true,\\"view\\":\\"home\\"},{\\"name\\":\\"my-page\\",\\"path\\":\\"/my-page\\",\\"view\\":\\"myPage\\"},{\\"name\\":\\"error\\",\\"path\\":\\"/error\\",\\"isDefault\\":true,\\"view\\":\\"error\\"}]);
    Talon.brandingService.setBrandingProperties({});
    
</script>
<script src=\\"/basePath/talon/view/latest/prod_debug/en_US/themeLayoutInner.js\\"></script>
<script src=\\"/basePath/talon/view/latest/prod_debug/en_US/myPage.js\\"></script>

<script type=\\"text/javascript\\" >
    Talon.componentService.createElement(\\"talon/app\\").then(function(element) {
        document.getElementById(\\"main\\").appendChild(element);
    });
</script>

Base path: /basePath
Version key: 1
Source nonce: "
`;

exports[`template-middleware generateHTML generates HTML in prod_debug_compat mode 1`] = `
"<script src=\\"/basePath/talon/framework/latest/prod_debug_compat/talon.js\\"></script>

    <script type=\\"text/javascript\\" >
    Talon.configProvider.register({
        getBasePath: function getBasePath() {
            return \\"/basePath\\";
        },
        getMode: function getMode() {
            return \\"prod_debug_compat\\";
        },
        getLocale: function getLocale() {
            return \\"en_US\\";
        },
        getUser: function getUser() {
            return {\\"isGuest\\":true};
        }
    });
    Talon.themeService.setTheme({\\"themeLayouts\\":{\\"inner\\":{\\"view\\":\\"themeLayoutInner\\"},\\"default\\":{\\"view\\":\\"themeLayoutDefault\\"}}});
    Talon.themeService.setViewToThemeLayoutMap({\\"home\\":\\"themeLayoutInner\\",\\"myPage\\":\\"themeLayoutInner\\",\\"error\\":\\"themeLayoutDefault\\"});
    Talon.routingService.registerRoutes([{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"isRoot\\":true,\\"view\\":\\"home\\"},{\\"name\\":\\"my-page\\",\\"path\\":\\"/my-page\\",\\"view\\":\\"myPage\\"},{\\"name\\":\\"error\\",\\"path\\":\\"/error\\",\\"isDefault\\":true,\\"view\\":\\"error\\"}]);
    Talon.brandingService.setBrandingProperties({});
    
</script>
<script src=\\"/basePath/talon/view/latest/prod_debug_compat/en_US/themeLayoutInner.js\\"></script>
<script src=\\"/basePath/talon/view/latest/prod_debug_compat/en_US/myPage.js\\"></script>

<script type=\\"text/javascript\\" >
    Talon.componentService.createElement(\\"talon/app\\").then(function(element) {
        document.getElementById(\\"main\\").appendChild(element);
    });
</script>

Base path: /basePath
Version key: 1
Source nonce: "
`;

exports[`template-middleware generateHTML generates HTML with script-src nonce 1`] = `
"<script src=\\"/basePath/talon/framework/latest/dev/talon.js\\"></script>

    <script type=\\"text/javascript\\" nonce=\\"T3stVa1ue\\">
    Talon.configProvider.register({
        getBasePath: function getBasePath() {
            return \\"/basePath\\";
        },
        getMode: function getMode() {
            return \\"dev\\";
        },
        getLocale: function getLocale() {
            return \\"en_US\\";
        },
        getUser: function getUser() {
            return {\\"isGuest\\":true};
        }
    });
    Talon.themeService.setTheme({\\"themeLayouts\\":{\\"inner\\":{\\"view\\":\\"themeLayoutInner\\"},\\"default\\":{\\"view\\":\\"themeLayoutDefault\\"}}});
    Talon.themeService.setViewToThemeLayoutMap({\\"home\\":\\"themeLayoutInner\\",\\"myPage\\":\\"themeLayoutInner\\",\\"error\\":\\"themeLayoutDefault\\"});
    Talon.routingService.registerRoutes([{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"isRoot\\":true,\\"view\\":\\"home\\"},{\\"name\\":\\"my-page\\",\\"path\\":\\"/my-page\\",\\"view\\":\\"myPage\\"},{\\"name\\":\\"error\\",\\"path\\":\\"/error\\",\\"isDefault\\":true,\\"view\\":\\"error\\"}]);
    Talon.brandingService.setBrandingProperties({});
    
</script>
<script src=\\"/basePath/talon/view/latest/dev/en_US/themeLayoutInner.js\\"></script>
<script src=\\"/basePath/talon/view/latest/dev/en_US/myPage.js\\"></script>

<script type=\\"text/javascript\\" nonce=\\"T3stVa1ue\\">
    Talon.componentService.createElement(\\"talon/app\\").then(function(element) {
        document.getElementById(\\"main\\").appendChild(element);
    });
</script>

Base path: /basePath
Version key: 1
Source nonce: T3stVa1ue"
`;

exports[`template-middleware generateHTML includes UIDs 1`] = `
"<script src=\\"/basePath/talon/framework/[dev uid for framework://talon]/dev/talon.js\\"></script>

    <script type=\\"text/javascript\\" >
    Talon.configProvider.register({
        getBasePath: function getBasePath() {
            return \\"/basePath\\";
        },
        getMode: function getMode() {
            return \\"dev\\";
        },
        getLocale: function getLocale() {
            return \\"en_US\\";
        },
        getUser: function getUser() {
            return {\\"isGuest\\":true};
        }
    });
    Talon.themeService.setTheme({\\"themeLayouts\\":{\\"inner\\":{\\"view\\":\\"themeLayoutInner\\"},\\"default\\":{\\"view\\":\\"themeLayoutDefault\\"}}});
    Talon.themeService.setViewToThemeLayoutMap({\\"home\\":\\"themeLayoutInner\\",\\"myPage\\":\\"themeLayoutInner\\",\\"error\\":\\"themeLayoutDefault\\"});
    Talon.routingService.registerRoutes([{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"isRoot\\":true,\\"view\\":\\"home\\"},{\\"name\\":\\"my-page\\",\\"path\\":\\"/my-page\\",\\"view\\":\\"myPage\\"},{\\"name\\":\\"error\\",\\"path\\":\\"/error\\",\\"isDefault\\":true,\\"view\\":\\"error\\"}]);
    Talon.brandingService.setBrandingProperties({});
    Talon.moduleRegistry.setResourceUids({\\"framework://talon\\":\\"[dev uid for framework://talon]\\",\\"view://home@en_US\\":\\"[dev uid for view://home@en_US]\\",\\"view://themeLayoutInner@en_US\\":\\"[dev uid for view://themeLayoutInner@en_US]\\"});
</script>
<script src=\\"/basePath/talon/view/[dev uid for view://themeLayoutInner@en_US]/dev/en_US/themeLayoutInner.js\\"></script>
<script src=\\"/basePath/talon/view/[dev uid for view://home@en_US]/dev/en_US/home.js\\"></script>

<script type=\\"text/javascript\\" >
    Talon.componentService.createElement(\\"talon/app\\").then(function(element) {
        document.getElementById(\\"main\\").appendChild(element);
    });
</script>

Base path: /basePath
Version key: 1
Source nonce: "
`;

exports[`template-middleware templateMiddleware send HTML 1`] = `
"<script src=\\"/basePath/talon/framework/[dev uid for framework://talon]/dev/talon.js\\"></script>

    <script type=\\"text/javascript\\" nonce=\\"T3stVa1ue\\">
    Talon.configProvider.register({
        getBasePath: function getBasePath() {
            return \\"/basePath\\";
        },
        getMode: function getMode() {
            return \\"dev\\";
        },
        getLocale: function getLocale() {
            return \\"en_US\\";
        },
        getUser: function getUser() {
            return {\\"isGuest\\":true};
        }
    });
    Talon.themeService.setTheme({\\"themeLayouts\\":{\\"inner\\":{\\"view\\":\\"themeLayoutInner\\"},\\"default\\":{\\"view\\":\\"themeLayoutDefault\\"}}});
    Talon.themeService.setViewToThemeLayoutMap({\\"home\\":\\"themeLayoutInner\\",\\"myPage\\":\\"themeLayoutInner\\",\\"error\\":\\"themeLayoutDefault\\"});
    Talon.routingService.registerRoutes([{\\"name\\":\\"home\\",\\"path\\":\\"/\\",\\"isRoot\\":true,\\"view\\":\\"home\\"},{\\"name\\":\\"my-page\\",\\"path\\":\\"/my-page\\",\\"view\\":\\"myPage\\"},{\\"name\\":\\"error\\",\\"path\\":\\"/error\\",\\"isDefault\\":true,\\"view\\":\\"error\\"}]);
    Talon.brandingService.setBrandingProperties({});
    Talon.moduleRegistry.setResourceUids({\\"framework://talon\\":\\"[dev uid for framework://talon]\\",\\"view://myPage@en_US\\":\\"[dev uid for view://myPage@en_US]\\",\\"view://themeLayoutInner@en_US\\":\\"[dev uid for view://themeLayoutInner@en_US]\\"});
</script>
<script src=\\"/basePath/talon/view/[dev uid for view://themeLayoutInner@en_US]/dev/en_US/themeLayoutInner.js\\"></script>
<script src=\\"/basePath/talon/view/[dev uid for view://myPage@en_US]/dev/en_US/myPage.js\\"></script>

<script type=\\"text/javascript\\" nonce=\\"T3stVa1ue\\">
    Talon.componentService.createElement(\\"talon/app\\").then(function(element) {
        document.getElementById(\\"main\\").appendChild(element);
    });
</script>

Base path: /basePath
Version key: 1
Source nonce: T3stVa1ue"
`;
